### PR TITLE
Added idrac_server_config_profile

### PR DESCRIPTION
**Pull request by jagadeeshnv**
##### SUMMARY
Submitting idrac_server_config_profile module for contribution

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
idrac_server_config_profile

##### ADDITIONAL INFORMATION

```
scp_status:
    {
      "Id": "JID_12345678",
      "JobState": "Completed",
      "JobType": "ImportConfiguration",
      "Message": "Successfully imported and applied Server Configuration Profile.",
      "MessageId": "SYS123",
      "Name": "Import Configuration",
      "PercentComplete": 100,
      "StartTime": "TIME_NOW",
      "Status": "Success"
    }
```

### CODE REVIEW

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    except RuntimeError as e:
+        module.fail_json(msg=str(e))
+    return import_status
+
+
+def run_export_server_config_profile(idrac, module):
+    """Export Server Configuration Profile to a network share."""
+    file_format = module.params['export_format'].lower()
+    export_format = ExportFormatEnum[module.params['export_format']]
+    scp_file_name_format = "%ip_%Y%m%d_%H%M%S_scp." + file_format
```
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: isn't it good to use format method for string append?

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    end_host_power_state = EndHostPowerStateEnum[module.params['end_host_power_state']]
+    shutdown_type = ShutdownTypeEnum[module.params['shutdown_type']]
+
+    idrac.use_redfish = True
+
+    try:
+        myshare = file_share_manager.create_share_obj(
+            share_path="{0}{1}{2}".format(share_name, os.sep, module.params['scp_file']),
+            creds=UserCredentials(module.params['share_user'],
+                                  module.params['share_pwd']), isFolder=False, )
```
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: do you need the ending comma?

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    job_wait = module.params['job_wait']
+
+    end_host_power_state = EndHostPowerStateEnum[module.params['end_host_power_state']]
+    shutdown_type = ShutdownTypeEnum[module.params['shutdown_type']]
+
+    idrac.use_redfish = True
+
+    try:
+        myshare = file_share_manager.create_share_obj(
+            share_path="{0}{1}{2}".format(share_name, os.sep, module.params['scp_file']),
```
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: do you really need this `share_name` variable, I can see you are using it only one place. You can use module option directly `module.params['share_name']` right?

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+            creds=UserCredentials(module.params['share_user'],
+                                  module.params['share_pwd']), isFolder=False, )
+        import_status = idrac.config_mgr.scp_import(myshare,
+                                                    target=target, shutdown_type=shutdown_type,
+                                                    end_host_power_state=end_host_power_state,
+                                                    job_wait=job_wait)
+        if not import_status or import_status.get('Status') != "Success":
+            module.fail_json(msg='Failed to import scp.', scp_status=import_status)
+
```
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: do you need this extra line

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+                                                      isFolder=True)
+        scp_file_name = myshare.new_file(scp_file_name_format)
+        export_status = idrac.config_mgr.scp_export(scp_file_name,
+                                                    target=target,
+                                                    export_format=export_format,
+                                                    export_use=export_use,
+                                                    job_wait=job_wait)
+        if not export_status or export_status.get('Status') != "Success":
+            module.fail_json(msg='Failed to export scp.', scp_status=export_status)
+
```
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: extra line

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+- name: Import Server Configuration Profile from a network share
+  dellemc_idrac_server_config_profile:
+    idrac_ip: "192.168.0.1"
+    idrac_user: "user_name"
+    idrac_pwd: "user_pwd"
+    command: "import"
+    share_name: "192.168.0.2:/share"
+    share_user: "share_user_name"
+    share_pwd: "share_user_pwd"
+    scp_file:   "scp_filename.xml"
```
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: normally one space after colon right?

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+- name: Import Server Configuration Profile from a local path
+  dellemc_idrac_server_config_profile:
+    idrac_ip: "192.168.0.1"
+    idrac_user: "user_name"
+    idrac_pwd: "user_pwd"
+    command: "import"
+    share_name: "/scp_folder"
+    share_user: "share_user_name"
+    share_pwd: "share_user_pwd"
+    scp_file:   "scp_filename.xml"
```
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: better to maintain single space after colon

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+                                                    target=target, shutdown_type=shutdown_type,
+                                                    end_host_power_state=end_host_power_state,
+                                                    job_wait=job_wait)
+        if not import_status or import_status.get('Status') != "Success":
+            module.fail_json(msg='Failed to import scp.', scp_status=import_status)
+    except RuntimeError as e:
+        module.fail_json(msg=str(e))
+    return import_status
+
+
```
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: Two line gap is required? applicable for similar places below as well

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+def run_export_server_config_profile(idrac, module):
+    """Export Server Configuration Profile to a network share."""
+    file_format = module.params['export_format'].lower()
```
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: Do you really need this `file_format` variable?

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+    export_format = ExportFormatEnum[module.params['export_format']]
+    scp_file_name_format = "%ip_%Y%m%d_%H%M%S_scp.{0}".format(file_format)
+    target = SCPTargetEnum[module.params['scp_components']]
+    job_wait = module.params['job_wait']
```
Comment (rajeevarakkal) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: can you pass `module.params['job_wait']` directly without using this temp variable ?

**Review by rajeevarakkal — APPROVED** [review]

**Comment by rajeevarakkal** [comment]
shipit

**Comment by jagadeeshnv** [comment]
bot_status

**Review by rajeevarakkal — APPROVED** [review]

**Comment by rajeevarakkal** [comment]
shipit

**Comment by jagadeeshnv** [comment]
bot_status

**Comment by jagadeeshnv** [comment]
ready_for_review

**Comment by jagadeeshnv** [comment]
@gundalow 
Kindly request you to review this new module?

**Comment by jagadeeshnv** [comment]
bot_status

**Comment by gundalow** [comment]
@rajeevarakkal Thank you for your review. One minor suggestion to make it easier for others to review is to click `Resolve Conversation` on the items raised once you've confirmed they have been fixed

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: dellemc_idrac_server_config_profile
+short_description: Export or Import Server Configuration Profile(SCP).
```
Comment (gundalow) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: I think this needs extending a bit, I'm not sure of the terminology here, though something like:
```suggestion
short_description: Export or Import Server Configuration Profile (SCP) on IDRAC.
```

The `short_description` is used on https://docs.ansible.com/ansible/latest/modules/list_of_all_modules.html

**Review by rajeevarakkal — APPROVED** [review]

**Comment by rajeevarakkal** [comment]
@gundalow , you are welcome, Sure..i will mark the resolved conversation appropriately.

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py (hunk 11 above)
Comment (jagadeeshnv) [reply lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: ok modified it.

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: dellemc_idrac_server_config_profile
```
Comment (jamescassell) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: Drop `dellecm_` prefix

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+options:
+  idrac_ip:
+    description: iDRAC IP Address.
+    type: str
+    required: True
+  idrac_user:
+    description: iDRAC username.
+    type: str
+    required: True
+  idrac_pwd:
```
Comment (jamescassell) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]:
```suggestion
  idrac_password:
```

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+    required: True
+  share_name:
+    description: CIFS or NFS Network Share or a local path.
+    type: str
+    required: True
+  share_user:
+    description: Network share user in the format 'user@domain' or 'domain\\user' if user is
+      part of a domain else 'user'. This option is mandatory for CIFS Network Share.
+    type: str
+  share_pwd:
```
Comment (jamescassell) [inline lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: pwd->password

**Review by jamescassell — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py (hunk 12 above)
Comment (jagadeeshnv) [reply lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: @jamescassell  Thank you for reviewing and your comments. This is under remote_management/dellemc/idrac/.  Also there is a previously contributed module in this path, hence the prefix. Anyway will update the PR as per your feedback

**Comment by jagadeeshnv** [comment]
bot_status

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py (hunk 13 above)
Comment (jagadeeshnv) [reply lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: @jamescassell We have this convention and also to maintain compatibility with our own previous release, we need the attribute name this way

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py (hunk 12 above)
Comment (jagadeeshnv) [reply lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: @jamescassell Done the renaming part

File: lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py (hunk 13 above)
Comment (jagadeeshnv) [reply lib/ansible/modules/remote_management/dellemc/idrac/dellemc_idrac_server_config_profile.py]: @jamescassell done changing pwd to password

**Comment by jagadeeshnv** [comment]
bot_status

**Review by jagadeeshnv — COMMENTED** [review]
@jamescassell  Have done the changes

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Thanks all, merged into `devel` for release in Ansible 2.8

**Pull request merged**